### PR TITLE
[Dynamo] Include UserDict in clone_inputs

### LINKED ISSUE #97724: `dynamo.utils.clone_inputs` failed with `BatchEncoding` from huggingface tokenizer
### 🐛 Describe the bug

```
from transformers import DistilBertTokenizer, DistilBertModel
text = "Replace me by any text you'd like."
tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
encoded_input = tokenizer(text, return_tensors="pt")

from torch._dynamo.utils import clone_inputs
clone_inputs(encoded_input) # this won't preserve the dict structure of the original inputs
```

### Versions

PyTorch version: 2.1.0a0+git7c52582
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.8.0 (default, Nov  6 2019, 21:49:08)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 510.108.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn.so.8.4.1
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.4.1
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.4.1
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.4.1
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.4.1
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.4.1
/usr/local/cuda-11.6/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          24
On-line CPU(s) list:             0-23
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           33
Model name:                      AMD Ryzen 9 5900X 12-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2200.000
CPU max MHz:                     4950.1948
CPU min MHz:                     2200.0000
BogoMIPS:                        7400.08
Virtualization:                  AMD-V
L1d cache:                       384 KiB
L1i cache:                       384 KiB
L2 cache:                        6 MiB
L3 cache:                        64 MiB
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.2
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.2.1
[pip3] functorch==1.14.0a0+408bcf1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.21.2
[pip3] pytorch-lightning==2.0.0
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.1
[pip3] torch==2.1.0a0+git7c52582
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.0.1+cu117
[pip3] torchdata==0.6.0
[pip3] torchmetrics==0.11.4
[pip3] torchrec-nightly==2023.3.16
[pip3] torchtext==0.15.1+cpu
[pip3] torchvision==0.15.1+cu117
[pip3] vector-quantize-pytorch==1.1.2
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.5.2                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.2.1                    pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] numpy                     1.21.2                   pypi_0    pypi
[conda] pytorch-lightning         2.0.0                    pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.1                    pypi_0    pypi
[conda] torch                     2.0.0+cu117              pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.0.1+cu117              pypi_0    pypi
[conda] torchdata                 0.6.0                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchrec-nightly          2023.3.16                pypi_0    pypi
[conda] torchtext                 0.15.1+cpu               pypi_0    pypi
[conda] torchvision               0.15.1+cu117             pypi_0    pypi
[conda] vector-quantize-pytorch   1.1.2                    pypi_0    pypi

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### PR DESCRIPTION
**Pull request by shingjan**
Fixes #97724 

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
Any chance of putting in a test? test/dynamo/test_repros.py is the usual spot

**Comment by shingjan** [comment]
@ezyang Thanks for the review. A test is added for `clone_inputs`. PTAL!

**Comment by ezyang** [comment]
Hmm... I'm not convinced we should fix this. `clone_inputs` is a private utility and not really intended for user use, and your test case shows that you only hit this case when you directly call this API. What are you trying to do with this API?

I'll also point out your API doesn't preserve the user dict, instead it turns into a plain dict after transform. Is that what you wanted?

Perhaps using `tree_map` with `clone_input` would be better?

**Review by jansel — CHANGES_REQUESTED** [review]
This doesn't seem exactly correct.

A user-define dict subclass could have arbitrary behavior that this wouldn't capture.  Dynamo should treat this more like a user defined object.

I think `clone_inputs` is also a bit wrong in that if one subclassed `dict` directly it would also happily copy that and remove the subclass.  It should be more like `type(example_inputs) is dict`

**Comment by shingjan** [comment]
Hi @ezyang and @jansel thanks for the review!

For the concern about `clone_inputs`, my use case is that I was running a bunch of huggingface pretrained models with its tokenizers. And the inputs generated from tokenizers are in a `BatchEncoding` class which inherit from `UserDict` instead of `dict` and acts essentially as a dict wrapper. `clone_inputs` will turn this subclass of `UserDict` to a list which is not acceptable by dynamo optimized modules of aforementioned huggingface models. So I thought it would be nice if `clone_inputs` can handle `BatchEncoding`s.  The only reason that a `dict` is called on `BatchEncoding` is that `same` cannot tell if a subclass object of `UserDict` and a dict object are the same. 

Based on Jason's point, I think it makes sense that if this `clone_inputs` are limited to internal use only and we should only convert it to a `dict` if it is a dict instead of a subclass of it. But if it also won't preserve a dict-like object for all dict subclasses like UserDict. If this should be something User should take care of, e.g. calling `dict(...)` explicitly on dict-like inputs, I am more than happy to make that change as well.

Thanks!

**Comment by jansel** [comment]
Re `OrderedDict`, the iteration order is the same as a normal dict: https://stackoverflow.com/questions/50872498/will-ordereddict-become-redundant-in-python-3-7

There is a tree_map in the private namespace `torch.utils._pytree`

> If this should be something User should take care of, e.g. calling dict(...) explicitly on dict-like inputs, I am more than happy to make that change as well.

I think this makes the most sense.

**Comment by shingjan** [comment]
@jansel The fix is in and I changed to test to make sure that an object of a subclass of `dict` will not be recognized as a `dict` and therefore `clone_inputs` will not preserve the dict from user.

**Review by yanboliang — APPROVED** [review]
